### PR TITLE
[GoogleDrive] Added support for captions (fixes issue #13619)

### PR DESCRIPTION
**Pull request by parmjitv**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x ] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [ x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [ x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ x] Bug fix
- [ ] Improvement
- [ ] New extractor
- [ ] New feature

---

Added support for video caption/subtitle files, VTT and TTML formats - issue #13619.

Cheers,

Parmjit V.

### CODE REVIEW

File: youtube_dl/extractor/googledrive.py [hunk 1]
```diff
@@ -97,10 +158,31 @@ def _real_extract(self, url):
...
 
+        hl = self._search_regex(
+            r'"hl"\s*,\s*"([^"]+)', webpage, 'hl')
+        ttsurl = self._search_regex(
+            r'"ttsurl"\s*,\s*"([^"]+)', webpage, 'ttsurl').encode('utf-8').decode('unicode_escape')
+        # the video Id for subtitles will be the last value in the ttsurl query string
+        video_subtitles_id = ttsurl.split('=')[-1]
+
+        subtitles = {}
+        subtitles_by_country = self._download_xml('https://drive.google.com/timedtext?id=%s&vid=%s&hl=%s&type=list&tlangs=1&v=%s&fmts=1&vssids=1' % (video_id, video_subtitles_id, hl, video_id), video_id)
```
Comment (dstftw) [inline youtube_dl/extractor/googledrive.py]: Subtitles requiring additional network requests should only be extracted when explicitly requested.

File: youtube_dl/extractor/googledrive.py [hunk 2]
```diff
@@ -24,7 +24,64 @@ class GoogleDriveIE(InfoExtractor):
...
-        'only_matching': True,
+        'md5': 'c230c67252874fddd8170e3fd1a45886',
+        'info_dict': {
+            'id': '1ENcQ_jeCuj7y19s66_Ou9dRP4GKGsodiDQ',
+            'ext': 'mp4',
+            'title': 'Andreea Banica feat Smiley - Hooky Song (Official Video).mp4',
+            'duration': 189,
+        },
+        'params': {
+            'only_matching': True,
```
Comment (dstftw) [inline youtube_dl/extractor/googledrive.py]: There is no such param `only_matching`.

File: youtube_dl/extractor/googledrive.py [hunk 3]
```diff
@@ -24,7 +24,64 @@ class GoogleDriveIE(InfoExtractor):
...
+        'md5': 'a5dd2666a9ee9da523b7a1a6f6c916ae',
+        'info_dict': {
+            'id': '0B_Cc_rAzzVX9blZZZzZJX2tNYVk',
+            'ext': 'mp4',
+            'title': 'Buffy 5x04 - Out of My Mind.mkv',
+            'duration': 2548,
+        },
+        'params': {
+            'only_matching': True,
+        }
```
Comment (dstftw) [inline youtube_dl/extractor/googledrive.py]: Remove all duplicate tests.

File: youtube_dl/extractor/googledrive.py [hunk 4]
```diff
@@ -97,10 +158,31 @@ def _real_extract(self, url):
             formats.append(f)
         self._sort_formats(formats)
 
+        hl = self._search_regex(
+            r'"hl"\s*,\s*"([^"]+)', webpage, 'hl')
+        ttsurl = self._search_regex(
+            r'"ttsurl"\s*,\s*"([^"]+)', webpage, 'ttsurl').encode('utf-8').decode('unicode_escape')
```
Comment (dstftw) [inline youtube_dl/extractor/googledrive.py]: Optional should not break extraction. Read FAQ.

**Comment by parmjitv** [comment]
Thanks for the prompt review.  I have made the requested corrections - please have a look at your convenience.  Cheers!

**Comment by parmjitv** [comment]
@dstftw - Does this PR require any further corrections?  Thank you!

**Comment by parmjitv** [comment]
@dstftw - Are any collaborators available for a second review of this PR? Thanks!

File: youtube_dl/extractor/googledrive.py [hunk 5]
```diff
@@ -53,6 +64,23 @@ def _extract_url(webpage):
...
+            return None
+        subtitles = {}
+        subtitles_by_country = self._download_xml('https://drive.google.com/timedtext?id=%s&vid=%s&hl=%s&type=list&tlangs=1&v=%s&fmts=1&vssids=1' % (video_id, video_subtitles_id, hl, video_id), video_id)
+        subtitle_available_tracks = subtitles_by_country.findall('track')
+        for subtitle_track in subtitle_available_tracks:
+            subtitle_lang_code = subtitle_track.attrib['lang_code']
+            subtitle_format_data = []
+            for subtitle_format in self._SUBTITLE_FORMATS_EXT:
+                subtitle_format_data.append({
+                    'url': 'https://drive.google.com/timedtext?vid=%s&v=%s&type=track&lang=%s&name&kind&fmt=%s' % (video_subtitles_id, video_id, subtitle_lang_code, subtitle_format),
```
Comment (dstftw) [inline youtube_dl/extractor/googledrive.py]: `update_url_query`.

File: youtube_dl/extractor/googledrive.py [hunk 6]
```diff
@@ -53,6 +64,23 @@ def _extract_url(webpage):
         if mobj:
             return 'https://drive.google.com/file/d/%s' % mobj.group('id')
 
+    def _get_subtitles(self, video_id, video_subtitles_id, hl):
```
Comment (dstftw) [inline youtube_dl/extractor/googledrive.py]: Automatic captions are also available.

Comment (parmjitv) [inline youtube_dl/extractor/googledrive.py]: Should automatic captions be considered/treated the same as subtitles?

Comment (dstftw) [inline youtube_dl/extractor/googledrive.py]: Automatic captions should be treated as automatic captions.

File: youtube_dl/extractor/googledrive.py [hunk 7]
```diff
@@ -44,6 +51,10 @@ class GoogleDriveIE(InfoExtractor):
         '46': 'webm',
         '59': 'mp4',
     }
+    _SUBTITLE_FORMATS_EXT = (
+        'vtt',
+        'ttml'
+    )
```
Comment (dstftw) [inline youtube_dl/extractor/googledrive.py]: How is this related to another extractors?

Comment (parmjitv) [inline youtube_dl/extractor/googledrive.py]: Whoops, never mind - I didn't read your comment properly :)

An updated PR will be submitted shortly.  Thanks for the follow up!

**Comment by parmjitv** [comment]
I have applied the latest fixes from the last review.  Please let me know if there is anything else to attend to for this PR.  Thank you!

**Comment by parmjitv** [comment]
@dstftw - Does this PR require any further corrections? Thank you!

File: youtube_dl/extractor/googledrive.py [hunk 8]
```diff
@@ -53,6 +64,98 @@ def _extract_url(webpage):
...
+                    'vid': video_subtitles_id,
+                    'v': video_id,
+                    'lang': subtitle_original_lang_code,
+                    'fmt': automatic_caption_format,
+                    'tlang': automatic_caption_lang_code,
+                    'name': '',
+                    'kind': '',
+                }
+                automatic_caption_format_data.append({
+                    'url': update_url_query('https://drive.google.com/timedtext?vid=%s&v=%s&type=track&lang=%s&name&kind&fmt=%s&tlang=%s', query),
```
Comment (dstftw) [inline youtube_dl/extractor/googledrive.py]: What's `vid=%s&v=%s&type=track&lang=%s&name&kind&fmt=%s&tlang=%s` for?

File: youtube_dl/extractor/googledrive.py [hunk 9]
```diff
@@ -53,6 +64,98 @@ def _extract_url(webpage):
...
+            if self._CAPTIONS_BY_COUNTRY_XML is None:
+                return None
+        self.to_screen('%s: Looking for automatic captions' % video_id)
+
+        subtitle_original_track = self._CAPTIONS_BY_COUNTRY_XML.find('track')
+        if subtitle_original_track is None:
+            return None
+        if not subtitle_original_track.attrib.get('lang_code'):
+            return None
+        subtitle_original_lang_code = subtitle_original_track.attrib['lang_code']
```
Comment (dstftw) [inline youtube_dl/extractor/googledrive.py]: Don't lookup `lang_code` twice.

File: youtube_dl/extractor/googledrive.py [hunk 10]
```diff
@@ -53,6 +64,98 @@ def _extract_url(webpage):
...
+
+        automatic_captions = {}
+        automatic_caption_available_targets = self._CAPTIONS_BY_COUNTRY_XML.findall('target')
+        for automatic_caption_target in automatic_caption_available_targets:
+            if not automatic_caption_target.attrib.get('lang_code'):
+                continue
+            automatic_caption_lang_code = automatic_caption_target.attrib['lang_code']
```
Comment (dstftw) [inline youtube_dl/extractor/googledrive.py]: Don't lookup twice.

File: youtube_dl/extractor/googledrive.py [hunk 11]
```diff
@@ -53,6 +64,98 @@ def _extract_url(webpage):
         if mobj:
             return 'https://drive.google.com/file/d/%s' % mobj.group('id')
 
+    def _set_captions_data(self, video_id, video_subtitles_id, hl):
+        try:
+            self._CAPTIONS_BY_COUNTRY_XML = self._download_xml(
```
Comment (dstftw) [inline youtube_dl/extractor/googledrive.py]: Caps is used for constants.

File: youtube_dl/extractor/googledrive.py [hunk 12]
```diff
@@ -53,6 +64,98 @@ def _extract_url(webpage):
...
+
+        subtitle_original_track = self._CAPTIONS_BY_COUNTRY_XML.find('track')
+        if subtitle_original_track is None:
+            return None
+        if not subtitle_original_track.attrib.get('lang_code'):
+            return None
+        subtitle_original_lang_code = subtitle_original_track.attrib['lang_code']
+
+        automatic_captions = {}
+        automatic_caption_available_targets = self._CAPTIONS_BY_COUNTRY_XML.findall('target')
```
Comment (dstftw) [inline youtube_dl/extractor/googledrive.py]: Move into loop.

Comment (parmjitv) [inline youtube_dl/extractor/googledrive.py]: Please clarify - the value of `subtitle_original_lang_code` will never change once it has been assigned, so what is the purpose of moving this into the `for` loop?

We need this value so that we can populate the query string for automatic captions later on in the method.

File: youtube_dl/extractor/googledrive.py (hunk 8 above)
Comment (parmjitv) [inline youtube_dl/extractor/googledrive.py]: As per the notes from issue #13619 - this is the query string that will yield the URL for automatic captions, for a given language code `tlang` and format `fmt` .

File: youtube_dl/extractor/googledrive.py (hunk 12 above)
Comment (dstftw) [inline youtube_dl/extractor/googledrive.py]: I did not suggest moving `subtitle_original_lang_code` into loop.

File: youtube_dl/extractor/googledrive.py (hunk 8 above)
Comment (dstftw) [inline youtube_dl/extractor/googledrive.py]: You already build query with `query` param.

File: youtube_dl/extractor/googledrive.py (hunk 12 above)
Comment (parmjitv) [inline youtube_dl/extractor/googledrive.py]: My apologies - were you referring to `automatic_caption_available_targets` ?

File: youtube_dl/extractor/googledrive.py (hunk 8 above)
Comment (parmjitv) [inline youtube_dl/extractor/googledrive.py]: Gotcha!  I will clean that up.

**Comment by parmjitv** [comment]
I have applied the latest fixes from the last review.  Please let me know if there is anything left outstanding.  Cheers!

**Comment by parmjitv** [comment]
@dstftw - Does this PR require any further corrections? Thank you!

**Comment by parmjitv** [comment]
@dstftw - Are any collaborators available for a further review of this PR? Thanks!

File: youtube_dl/extractor/googledrive.py [hunk 13]
```diff
@@ -53,6 +64,98 @@ def _extract_url(webpage):
         if mobj:
             return 'https://drive.google.com/file/d/%s' % mobj.group('id')
 
+    def _set_captions_data(self, video_id, video_subtitles_id, hl):
+        try:
+            self._captions_by_country_xml = self._download_xml(
+                'https://drive.google.com/timedtext?id=%s&vid=%s&hl=%s&type=list&tlangs=1&v=%s&fmts=1&vssids=1', video_id, query={
```
Comment (dstftw) [inline youtube_dl/extractor/googledrive.py]: `?id=%s&vid=%s&hl=%s&type=list&tlangs=1&v=%s&fmts=1&vssids=1`.

File: youtube_dl/extractor/googledrive.py [hunk 14]
```diff
@@ -53,6 +64,98 @@ def _extract_url(webpage):
...
+                    'kind': '',
+                }
+                automatic_caption_format_data.append({
+                    'url': update_url_query('https://drive.google.com/timedtext', query),
+                    'ext': automatic_caption_format,
+                })
+            automatic_captions[automatic_caption_lang_code] = automatic_caption_format_data
+        if not automatic_captions:
+            self.report_warning('video doesn\'t have automatic captions')
+        return automatic_captions
```
Comment (dstftw) [inline youtube_dl/extractor/googledrive.py]: 92-115, 133-157 codes are identical apart from entry tag and query.

**Comment by parmjitv** [comment]
@dstftw - Fixed, and fixed.  Thanks for pointing those out.  Shall we apply any further corrections to this PR?

**Comment by parmjitv** [comment]
@dstftw - Does this PR require any further corrections? Thank you!

**Comment by parmjitv** [comment]
@dstftw - Are any collaborators available for a further review of this PR? Thanks!

**Comment by parmjitv** [comment]
@dstftw - Would it be possible to have another review of this PR?  I believe we have resolved the last of the outstanding issues.  Much appreciated!
